partitionccl: refactor implicit columns detection to CCL

otan opened the pull request:
This puts the CCL features with the associated code, and determines the
implicit partitioning as part of the create partitioning function, instead
of requiring it to be passed in.

Also added some unimplemented errors to bits we have not done yet.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/59462)
<!-- Reviewable:end -->

ajstorm reviewed (approved):
Changes <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

Nit about the commit record - I personally find the "and makes create partitioning require two functions" part confusing.  Isn't this change bring the two distinct function calls under a single call (thus hiding the need for two)?  If so, we might want to reword to "...and determines the implicit partitioning as part of the create partitioning function, instead of requiring it to be passed in."

> Reviewed 1 of 7 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59462#-:-MS5X8gbEJ_6OP5XqsIq:bvdyimu)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/alter_index.go, line 95 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59462#-MS82OnX0qblHE79Ta53:-MS82OnX0qblHE79Ta54:b-2pu1b1) ([raw file](https://github.com/cockroachdb/cockroach/blob/eeb0e3ff4f29b708e5a6f359a15df249473f69b4/pkg/sql/alter_index.go#L95)):*
> ```Go
> 				return unimplemented.New(
> 					"ALTER INDEX PARTITION BY",
> 					"cannot ALTER INDEX PARTITION BY on an index which creates implicit partitioning",
> ```

A couple of nits:

- Instead of saying "cannot ALTER INDEX PARTITION BY" it might be clearer if we say "cannot ALTER and index and change its partitioning..."  
- I think this one would benefit from a "Hint".  Implicit partitioning, by definition, happens without the user's explicit instruction (for some definition of explicit).  As a result, I'd anticipate that many won't know it's going on, and as a result, this error message will be confusing.  A doc link could also help, once one's available.

Both of these comments apply to this message, as well as the one above (from a previous PR).
___
*[pkg/sql/alter_table.go, line 798 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59462#-MS83Zqy0H7AxcYXZVoa:-MS83Zqy0H7AxcYXZVob:b-3w1gvj) ([raw file](https://github.com/cockroachdb/cockroach/blob/eeb0e3ff4f29b708e5a6f359a15df249473f69b4/pkg/sql/alter_table.go#L798)):*
> ```Go
> 				return unimplemented.New(
> 					"ALTER INDEX PARTITION BY",
> 					"cannot ALTER INDEX PARTITION BY on an index which creates implicit partitioning",
> ```

Same comment as above.

Additionally, is this error message correct?  We're in ALTER TABLE modifying the primary key, right?
___
*[pkg/sql/create_table.go, line 1155 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59462#-MS898yGDc_Yu4KKyoHa:-MS898yGDc_Yu4KKyoHb:b-5ihquc) ([raw file](https://github.com/cockroachdb/cockroach/blob/eeb0e3ff4f29b708e5a6f359a15df249473f69b4/pkg/sql/create_table.go#L1155)):*
> ```Go
> 
> // CreatePartitioning returns a new index descriptor with partitioning fields
> // populated to align with the partBy clause.
> ```

Nit: Might be better to say PartitionBy clause here.  partBy is just the name of the variable.
___
*[pkg/sql/create_table.go, line 1168 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59462#-MS89O9x05tnqZuA6YAl:-MS89O9x05tnqZuA6YAm:blt3hjn) ([raw file](https://github.com/cockroachdb/cockroach/blob/eeb0e3ff4f29b708e5a6f359a15df249473f69b4/pkg/sql/create_table.go#L1168)):*
> ```Go
> 			return indexDesc, unimplemented.Newf(
> 				"ALTER ... PARTITION BY NOTHING",
> 				"cannot PARTITION BY NOTHING if the partition is implicitly partitioned",
> ```

Nit: _object_ is implicitly partitioned?  Can a partition be partitioned :exploding_head: 

Also, we might want to say "cannot _ALTER_ to PARTITION BY NOTHING".


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59462)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajstorm)

___
*[pkg/sql/alter_index.go, line 95 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59462#-MS82OnX0qblHE79Ta53:-MS9RfHTDAbwrubcyMD1:b-7n5co2) ([raw file](https://github.com/cockroachdb/cockroach/blob/eeb0e3ff4f29b708e5a6f359a15df249473f69b4/pkg/sql/alter_index.go#L95)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

A couple of nits:

- Instead of saying "cannot ALTER INDEX PARTITION BY" it might be clearer if we say "cannot ALTER and index and change its partitioning..."  
- I think this one would benefit from a "Hint".  Implicit partitioning, by definition, happens without the user's explicit instruction (for some definition of explicit).  As a result, I'd anticipate that many won't know it's going on, and as a result, this error message will be confusing.  A doc link could also help, once one's available.

Both of these comments apply to this message, as well as the one above (from a previous PR).
</blockquote></details>

It is gated by an experimental flag so in that way it seems explicit.
When the gate is removed, this message will not occur, since it's likely to have been implemented.
___
*[pkg/sql/alter_table.go, line 798 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59462#-MS83Zqy0H7AxcYXZVoa:-MS9Ru2G3I_nZbSgfgee:b-yolr8q) ([raw file](https://github.com/cockroachdb/cockroach/blob/eeb0e3ff4f29b708e5a6f359a15df249473f69b4/pkg/sql/alter_table.go#L798)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Same comment as above.

Additionally, is this error message correct?  We're in ALTER TABLE modifying the primary key, right?
</blockquote></details>

Done, except for a the hint.


<!-- Sent from Reviewable.io -->

otan committed:
partitionccl: refactor implicit columns detection to CCL

This puts the CCL features with the associated code, and determines the
implicit partitioning as part of the create partitioning function, instead
of requiring it to be passed in.

Also added some unimplemented errors to bits we have not done yet.

Release note: None

otan commented:
bors r=ajstorm
